File: examples/example_txth_writer.cpp
//
// Copyright (c) 2026, Bayerische Motoren Werke Aktiengesellschaft (BMW AG)
// SPDX-License-Identifier: MPL-2.0
//
/**
 * \file
 * \brief Write an OSI `.txth` text trace file.
 */

#include <osi-utilities/tracefile/writer/TXTHTraceFileWriter.h>

#include <filesystem>

#include "osi_sensordata.pb.h"
#include "osi_version.pb.h"

/**
 * \brief Get the current time in OSI filename format.
 * \return Timestamp formatted as YYYYMMDDThhmmssZ.
 */
std::string GetCurrentTimeAsString() {
    const auto now = std::chrono::system_clock::now();
    const auto timer = std::chrono::system_clock::to_time_t(now);
    const std::tm utc_time_structure = *std::gmtime(&timer);  // Greenwich Mean Time (GMT) is in Coordinated Universal Time (UTC) zone
    std::ostringstream oss;
    oss << std::put_time(&utc_time_structure, "%Y%m%dT%H%M%S");
    oss << "Z";  // As GMT is used as a reference time zone, add Z to indicate UTC (+00:00)
    return oss.str();
}

/**
 * \brief Generate a temporary output path for the `.txth` example.
 * \return Temporary file path.
 */
std::filesystem::path GenerateTempFilePath() {
    // create a path which follows the OSI specification recommendation
    std::string file_name = GetCurrentTimeAsString();
    const auto osi_version = osi3::InterfaceVersion::descriptor()->file()->options().GetExtension(osi3::current_interface_version);
    file_name += "_" + std::to_string(osi_version.version_major()) + "." + std::to_string(osi_version.version_minor()) + "." + std::to_string(osi_version.version_patch());
    file_name += "_" + google::protobuf::internal::VersionString(GOOGLE_PROTOBUF_VERSION);
    file_name += "_10";  // 10 frames
    file_name += "_example-txth-writer.txth";
    auto path = std::filesystem::temp_directory_path() / file_name;
    return path;
}

/**
 * \brief Entry point for the TXTH writer example.
 */
int main(int argc, const char** argv) {
    std::cout << "Starting txth writer example:" << std::endl;

    // Create writer and open file
    auto trace_file_writer = osi3::TXTHTraceFileWriter();
    const auto trace_file_path = GenerateTempFilePath();
    std::cout << "Creating trace_file at " << trace_file_path << std::endl;
    trace_file_writer.Open(trace_file_path);

    // create OSI data to store
    const auto osi_version = osi3::InterfaceVersion::descriptor()->file()->options().GetExtension(osi3::current_interface_version);

    osi3::SensorView sensor_view;
    sensor_view.mutable_version()->CopyFrom(osi_version);
    sensor_view.mutable_sensor_id()->set_value(0);

    auto* const ground_truth = sensor_view.mutable_global_ground_truth();
    ground_truth->mutable_version()->CopyFrom(osi_version);

    auto* const host_vehicle = ground_truth->mutable_moving_object()->Add();
    host_vehicle->mutable_id()->set_value(12);
    host_vehicle->mutable_vehicle_classification()->set_type(osi3::MovingObject_VehicleClassification_Type_TYPE_SMALL_CAR);
    host_vehicle->mutable_base()->mutable_dimension()->set_length(5);
    host_vehicle->mutable_base()->mutable_dimension()->set_width(2);
    host_vehicle->mutable_base()->mutable_dimension()->set_height(1.5);
    host_vehicle->mutable_base()->mutable_velocity()->set_x(10.0);

    // write the data continuously in a loop
    constexpr double kTimeStepSizeS = 0.1;  // NOLINT
    for (int i = 0; i < 10; ++i) {
        // manipulate the data so not every message is the same
        auto timestamp = sensor_view.timestamp().seconds() * 1000000000 + sensor_view.timestamp().nanos();
        timestamp += kTimeStepSizeS * 1000000000;
        sensor_view.mutable_timestamp()->set_nanos(timestamp % 1000000000);
        sensor_view.mutable_timestamp()->set_seconds(timestamp / 1000000000);
        ground_truth->mutable_timestamp()->set_nanos(timestamp % 1000000000);
        ground_truth->mutable_timestamp()->set_seconds(timestamp / 1000000000);
        const auto old_position = host_vehicle->base().position().x();
        const auto new_position = old_position + host_vehicle->base().velocity().x() * kTimeStepSizeS;
        host_vehicle->mutable_base()->mutable_position()->set_x(new_position);

        // write the data
        trace_file_writer.WriteMessage(sensor_view);
    }

    trace_file_writer.Close();

    std::cout << "Finished txth writer example" << std::endl;
    return 0;
}

File: python/examples/example_txth_writer.py
# SPDX-License-Identifier: MPL-2.0
# SPDX-FileCopyrightText: Copyright (c) 2026, Bayerische Motoren Werke Aktiengesellschaft (BMW AG)

"""Write an OSI .txth text trace file."""

from __future__ import annotations

import os
import sys
from datetime import datetime, timezone
from pathlib import Path

import google.protobuf
from osi3.osi_sensorview_pb2 import SensorView

from osi_utilities import TXTHTraceFileWriter
from osi_utilities.tracefile.timestamp import timestamp_to_nanoseconds


def _generate_osi_filename(description: str, extension: str, frame_count: int = 10) -> str:
    """Generate filename following OSI naming convention."""
    now = datetime.now(timezone.utc).strftime("%Y%m%dT%H%M%SZ")
    try:
        from osi3.osi_version_pb2 import DESCRIPTOR as VERSION_DESCRIPTOR
        from osi3.osi_version_pb2 import InterfaceVersion

        osi_ver = VERSION_DESCRIPTOR.file_options.Extensions[InterfaceVersion.current_interface_version]
        osi_version = f"{osi_ver.version_major}.{osi_ver.version_minor}.{osi_ver.version_patch}"
    except Exception:
        osi_version = "0.0.0"
    proto_version = google.protobuf.__version__
    pid = os.getpid()
    return f"{now}_sv_{osi_version}_{proto_version}_{frame_count}_{pid}_{description}.{extension}"


def main() -> int:
    """Create a .txth file with 10 SensorView frames in text format."""
    print("Starting TXTH writer example:")

    file_name = _generate_osi_filename("example_txth_writer", "txth")
    output_dir = Path(__file__).resolve().parent.parent.parent / ".playground"
    output_dir.mkdir(exist_ok=True)
    trace_file_path = output_dir / file_name
    print(f"Creating trace file at {trace_file_path}")

    writer = TXTHTraceFileWriter()
    if not writer.open(trace_file_path):
        print(f"Error: Could not open '{trace_file_path}'", file=sys.stderr)
        return 1

    # Build a SensorView with one moving object
    sensor_view = SensorView()
    sensor_view.sensor_id.value = 0
    gt = sensor_view.global_ground_truth
    host = gt.moving_object.add()
    host.id.value = 12
    host.vehicle_classification.type = 2  # TYPE_SMALL_CAR
    host.base.dimension.length = 5
    host.base.dimension.width = 2
    host.base.dimension.height = 1.5
    host.base.velocity.x = 10.0

    time_step_ns = int(0.1 * 1e9)
    with writer:
        for i in range(10):
            ts_ns = timestamp_to_nanoseconds(sensor_view) + time_step_ns
            sensor_view.timestamp.seconds = int(ts_ns // 1e9)
            sensor_view.timestamp.nanos = int(ts_ns % 1e9)
            gt.timestamp.CopyFrom(sensor_view.timestamp)

            host.base.position.x += host.base.velocity.x * 0.1

            if not writer.write_message(sensor_view):
                print(f"Error: Could not write message {i}", file=sys.stderr)
                return 1

    print("Finished TXTH writer example")
    return 0


if __name__ == "__main__":
    sys.exit(main())
